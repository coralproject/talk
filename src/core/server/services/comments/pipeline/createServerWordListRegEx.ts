import { defaults } from "lodash";
import RE2 from "re2";

import { LanguageCode } from "coral-common/helpers";
import { DeepPartial } from "coral-common/types";

export interface WordListRule {
  boundary: string;
  punctuation: string;
}

export const DefaultWordListRule: WordListRule = {
  // The following symbol, \p{L} refers to any letter class within unicode.
  // Because we're adding the ^, we're also saying to exclude any from that set,
  // leaving all non-word characters from unicode available for selection.
  boundary: "[^\\p{L}]+",
  punctuation: "[\\s\"'?!.,¿¡`:;]+",
};

export const WordListRules: DeepPartial<Record<LanguageCode, WordListRule>> = {
  "en-US": DefaultWordListRule,
};

/**
 * Escape string for special regular expression characters.
 *
 * @param str the string to escape from regex characters
 */
export function escapeRegExp(str: string) {
  // $& means the whole matched string
  return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/**
 * generateRegExp will generate the tester that can be used to test strings
 * for matches on phrases.
 *
 * @param lang the language to possibly swap word list rules
 * @param phrases the phrases to use for creating the expression
 */
export default function createServerWordListRegEx(
  lang: LanguageCode,
  phrases: string[]
) {
  // Get the rule list for this language, fallback to english if we haven't
  // provided any overrides.
  const rule: WordListRule = defaults(
    WordListRules[lang] || {},
    DefaultWordListRule
  );

  // Split up the words from the list into a regex escaped string.
  const words = phrases
    .map((phrase) =>
      phrase
        // Split each phrase by whitespace.
        .split(/\s/)
        // Escape each phrase, we don't expect any of them to contain regex.
        .map((word) => escapeRegExp(word))
        // Rejoin to ensure that any variation of the word separated by a
        // punctuation character should also be caught.
        .join(rule.punctuation)
    )
    // For each of these words, wrap a `|` or OR.
    .join("|");

  // Wrap the pattern in split rules. We want to match any word that either is
  // at the start of a string, or a word boundary. The word must also either be
  // at the end of the string or at another word boundary.
  const pattern = `(^|${rule.boundary})(${words})($|${rule.boundary})`;

  return new RE2(pattern, "iuA");
}
