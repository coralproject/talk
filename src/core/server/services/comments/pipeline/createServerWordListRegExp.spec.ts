import RE2 from "re2";

import createServerWordListRegEx from "coral-server/services/comments/pipeline/createServerWordListRegEx";

const buildTester = (re: RE2) => (str: string) => re.test(str);

const buildSplitter = (re: RE2) => (str: string) => str.split(re);

describe("en-US", () => {
  const re = createServerWordListRegEx("en-US", [
    "bad",
    "french fries",
    "worse",
    "jalapeño",
    "1km",
  ]);

  const test = buildTester(re);

  it("test words in the list", () => {
    expect(test("bad")).toBeTruthy();
    expect(test("worse")).toBeTruthy();
    expect(test("1km")).toBeTruthy();
  });

  it("test repeated words in the list", () => {
    expect(test("bad bad bad")).toBeTruthy();
    expect(test("worse worse worse")).toBeTruthy();
  });

  it("test words not in the list", () => {
    expect(test("fine")).toBeFalsy();
    expect(test("ok")).toBeFalsy();
  });

  it("test words that end with a unicode character", () => {
    expect(test("I have one jalapeño.")).toBeTruthy();
    expect(test("I have two jalapeños.")).toBeFalsy();
  });

  it("test words in the list while being case insensitive", () => {
    expect(test("Bad")).toBeTruthy();
    expect(test("BAd")).toBeTruthy();
    expect(test("BAD")).toBeTruthy();
    expect(test("bAD")).toBeTruthy();
    expect(test("baD")).toBeTruthy();
    expect(test("bAd")).toBeTruthy();
    expect(test("1KM")).toBeTruthy();
  });

  it("test multi-words", () => {
    expect(test("french fries")).toBeTruthy();
    expect(test("french!fries")).toBeTruthy();
    expect(test("french.fries")).toBeTruthy();
    expect(test("french?fries")).toBeTruthy();
    expect(test("french¿fries")).toBeTruthy();
    expect(test("french:fries")).toBeTruthy();
    expect(test("french;fries")).toBeTruthy();
  });

  it("test words at the end of a sentence", () => {
    expect(test("this sentence is bad.")).toBeTruthy();
    expect(test("this sentence is worse.")).toBeTruthy();
    expect(test("this sentence has french fries.")).toBeTruthy();
    expect(test("this sentence has french!fries.")).toBeTruthy();
    expect(test("this sentence has french.fries.")).toBeTruthy();
    expect(test("this sentence has french?fries.")).toBeTruthy();
    expect(test("this sentence has french¿fries.")).toBeTruthy();
    expect(test("this sentence has french:fries.")).toBeTruthy();
    expect(test("this sentence has french;fries.")).toBeTruthy();
  });

  it("test words at the start of a sentence", () => {
    expect(test("bad is the start of the sentence.")).toBeTruthy();
    expect(test("worse is the start of the sentence.")).toBeTruthy();
    expect(test("french fries is the start of the sentence.")).toBeTruthy();
  });

  it("test does not preserve state", () => {
    expect(test("bad 1")).toBeTruthy();
    expect(test("bad 2")).toBeTruthy();
    expect(test("more bad 3")).toBeTruthy();
    expect(test("more bad 4")).toBeTruthy();
  });

  it("test repeated words", () => {
    expect(test("This is bad bad, very bad")).toBeTruthy();
  });

  it("test does not match substrings", () => {
    expect(test("baddd")).toBeFalsy();
    expect(test("wwworse")).toBeFalsy();
    expect(test("fffrench fries")).toBeFalsy();
  });

  it("test handles when there are numbers", () => {
    expect(test("bad3")).toBeTruthy();
    expect(test("3bad")).toBeTruthy();
  });

  const split = buildSplitter(re);

  it("splits the words in a sentence correctly", () => {
    expect(split("this sentence is bad.")).toMatchInlineSnapshot(`
      Array [
        "this sentence is",
        " ",
        "bad",
        ".",
        "",
      ]
    `);
    expect(split("this sentence is worse.")).toMatchInlineSnapshot(`
      Array [
        "this sentence is",
        " ",
        "worse",
        ".",
        "",
      ]
    `);
  });

  it("splits words when there are repeat words", () => {
    expect(split("This is bad bad, very BAD.")).toMatchInlineSnapshot(`
      Array [
        "This is",
        " ",
        "bad",
        " ",
        "bad, very",
        " ",
        "BAD",
        ".",
        "",
      ]
    `);
  });

  it("splits the words with unicode in a sentence correctly", () => {
    expect(split("this sentence has one jalapeño.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has one",
        " ",
        "jalapeño",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has many jalapeños.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has many jalapeños.",
      ]
    `);
  });

  it("splits the multi-words in a sentence correctly", () => {
    expect(split("this sentence has french fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french fries",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has french;fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french;fries",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has french!fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french!fries",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has french.fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french.fries",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has french?fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french?fries",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has french¿fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french¿fries",
        ".",
        "",
      ]
    `);
    expect(split("this sentence has french:fries.")).toMatchInlineSnapshot(`
      Array [
        "this sentence has",
        " ",
        "french:fries",
        ".",
        "",
      ]
    `);
  });
});

describe("es", () => {
  const re = createServerWordListRegEx("es", ["adónde vas", "tú"]);

  const test = buildTester(re);

  it("test words in the list", () => {
    expect(test("Pablo, ¿adónde vas?")).toBeTruthy();
    expect(test("Estoy cansado, ¿y tú?")).toBeTruthy();
    expect(test("¿tú?")).toBeTruthy();
  });
});

describe("pt-BR", () => {
  const re = createServerWordListRegEx("pt-BR", [
    "bi",
    "outro",
    "café",
    "m e r d a",
    "Como fazer coisas ruins",
  ]);

  const test = buildTester(re);

  it("test words in the list", () => {
    expect(test("biólogo se soletra com: bi ")).toBeTruthy();
    expect(test("m.e.r.d.a")).toBeTruthy();
    expect(test("não tomo café pois faz mal")).toBeTruthy();
    expect(test("Como fazer coisas ruins")).toBeTruthy();
  });

  it("test words not in the list", () => {
    expect(test("O biólogo recomenda este artigo")).toBeFalsy();
    expect(test("cafe")).toBeFalsy();
    expect(test("Ser banido é uma merda")).toBeFalsy();
  });

  const split = buildSplitter(re);

  it("splits the words with unicode in a sentence correctly", () => {
    expect(split("biólogo se soletra com: bi ")).toMatchInlineSnapshot(`
      Array [
        "biólogo se soletra com",
        ": ",
        "bi",
        " ",
        "",
      ]
    `);
    expect(split("m.e.r.d.a")).toMatchInlineSnapshot(`
      Array [
        "",
        "",
        "m.e.r.d.a",
        "",
        "",
      ]
    `);
    expect(split("não tomo café pois faz mal")).toMatchInlineSnapshot(`
      Array [
        "não tomo",
        " ",
        "café",
        " ",
        "pois faz mal",
      ]
    `);
    expect(split("Como fazer coisas ruins")).toMatchInlineSnapshot(`
      Array [
        "",
        "",
        "Como fazer coisas ruins",
        "",
        "",
      ]
    `);
    expect(split("O biólogo recomenda este artigo")).toMatchInlineSnapshot(`
      Array [
        "O biólogo recomenda este artigo",
      ]
    `);
    expect(split("cafe")).toMatchInlineSnapshot(`
      Array [
        "cafe",
      ]
    `);
    expect(split("Ser banido é uma merda")).toMatchInlineSnapshot(`
      Array [
        "Ser banido é uma merda",
      ]
    `);
  });
});
