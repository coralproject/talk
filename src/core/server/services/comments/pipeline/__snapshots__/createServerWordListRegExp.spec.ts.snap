// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`en-US splits the multi-words in a sentence correctly 1`] = `
Array [
  "this sentence has",
  " ",
  "french fries",
  ".",
  "",
]
`;

exports[`en-US splits the multi-words in a sentence correctly 2`] = `
Array [
  "this sentence has",
  " ",
  "french;fries",
  ".",
  "",
]
`;

exports[`en-US splits the multi-words in a sentence correctly 3`] = `
Array [
  "this sentence has",
  " ",
  "french!fries",
  ".",
  "",
]
`;

exports[`en-US splits the multi-words in a sentence correctly 4`] = `
Array [
  "this sentence has",
  " ",
  "french.fries",
  ".",
  "",
]
`;

exports[`en-US splits the multi-words in a sentence correctly 5`] = `
Array [
  "this sentence has",
  " ",
  "french?fries",
  ".",
  "",
]
`;

exports[`en-US splits the multi-words in a sentence correctly 6`] = `
Array [
  "this sentence has",
  " ",
  "french¿fries",
  ".",
  "",
]
`;

exports[`en-US splits the multi-words in a sentence correctly 7`] = `
Array [
  "this sentence has",
  " ",
  "french:fries",
  ".",
  "",
]
`;

exports[`en-US splits the words in a sentence correctly 1`] = `
Array [
  "this sentence is",
  " ",
  "bad",
  ".",
  "",
]
`;

exports[`en-US splits the words in a sentence correctly 2`] = `
Array [
  "this sentence is",
  " ",
  "worse",
  ".",
  "",
]
`;

exports[`en-US splits the words with unicode in a sentence correctly 1`] = `
Array [
  "this sentence has one",
  " ",
  "jalapeño",
  ".",
  "",
]
`;

exports[`en-US splits the words with unicode in a sentence correctly 2`] = `
Array [
  "this sentence has many jalapeños.",
]
`;

exports[`en-US splits words when there are repeat words 1`] = `
Array [
  "This is",
  " ",
  "bad",
  " ",
  "bad, very BAD.",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 1`] = `
Array [
  "biólogo se soletra com",
  ": ",
  "bi",
  " ",
  "",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 2`] = `
Array [
  "",
  "",
  "m.e.r.d.a",
  "",
  "",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 3`] = `
Array [
  "não tomo",
  " ",
  "café",
  " ",
  "pois faz mal",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 4`] = `
Array [
  "",
  "",
  "Como fazer coisas ruins",
  "",
  "",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 5`] = `
Array [
  "O biólogo recomenda este artigo",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 6`] = `
Array [
  "cafe",
]
`;

exports[`pt-BR splits the words with unicode in a sentence correctly 7`] = `
Array [
  "Ser banido é uma merda",
]
`;
